jax/interpreters/ad.py

```diff
@@ -315,34 +315,33 @@ def process_primitive(self, primitive, tracers, params):
   def process_call(self, call_primitive, f, tracers, params):
     assert call_primitive.multiple_results
     primals, tangents = unzip2((t.primal, t.tangent) for t in tracers)
-    nonzero_tangents, tangent_tree_def = tree_flatten(tangents)
-    nz_tangents = [type(t) is not Zero for t in tangents]
+    which_nz = [     type(t) is not Zero           for t in tangents]
```

Comment:
The `tangents` here contain symbolic zeros (i.e. `Zero` instances), which because they are not valid jaxtypes must be squeezed out before passed to the call primitive bind. The `traceable` wrapper applied below would ensure they were put back before the primals and tangents were put back together into `JVPTracer`s and dropped into the function being called.

However, that led to a caching issue: each `Zero` instance carries with it an aval, so in the case where we were applying a function with input type `[(i32[], False), (f32[DBIdx(0)], True)]` to the explicit primal argument `f32[N]` with some accompanying `Zero(f32[N])` for some literal value `N`, the pytree metadata would be specialized on a particular value of `N` and hence we wouldn't get cache hits for different values of `N` as we intend.

Actually, the issue was a bit worse: the `Zero`'s aval wouldn't be correct under the call, because it would need to agree with the primal aval of `f32[n]`! Something analogous happens on the output side too.

So it seems better not to propagate `Zero`s' avals through calls. Instead, we just need to propagate where the zeros are (e.g. using Nones, so we can still use the pytree machinery). We can construct the correct agrees-with-the-primal aval on the other side.

So that's what these changes do!